### ISSUE TITLE
Refactor insertion markers to not be blocks

### ISSUE THREAD

**Issue by BeksOmega** [google/blockly#5375]
**Is your feature request related to a problem? Please describe.**

Insertion markers cause lots of bugs in Blockly. This is because they are just blocks with an extra flag set to true, instead of being their own independent type. As such, anywhere that deals with blocks has to explicitly filter out insertion markers, and if we forget, we get bugs.

Here is a list of bugs related to this problem with insertion markers:
  * #2216
  * #2249
  * #2322
  * #2354
  * #2948
  * #4048
  * #4485 
  * #4636 
  * #5374 

There may be more that I didn't catch.

**Describe the solution you'd like**

Refactor the insertion markers to not be blocks. They should be their own type that does not come up in getDescendants or getChildren or getAllBlocks calls. They should have connections (or some connection-like interface) but they should not have fields or icons.

Basically, we should use the type system to make sure these errors never occur in the first place, instead of checking flags and chasing bugs.

**Issues**
One big issue with this would be figuring out how to make the insertion marker have the correct shape. Right now it works by:
  1) Creating a block
  2) Setting all of the field values so they have the correct widths
  3) Rendering the block (since the rendering system is tightly coupled to the BlockSvg type)
  4) Hiding all of the fields so the user can't see them
  5) Setting the block's color to grey so it looks like an "insertion marker" (even though it's just a block with a flag)

The rendering system would need to be decoupled from the BlockSvg type before we can refactor insertion markers.
